In [ ]:
# Problem 1:

## Dependencies
import numpy as np

## Data: yearly salary in ($1000) [2017, 2018, 2019]
alice = [99, 101, 103]
bob = [110, 108, 105]
tim = [90, 88, 85]

salaries = np.array([alice, bob, tim])
taxation = np.array([[0.2, 0.25, 0.22],
[0.4, 0.5, 0.5],
[0.1, 0.2, 0.1]])


## One-liner
max_income = np.max(salaries - salaries * taxation)

## Result
print(max_income)

81.0


In [ ]:
# Problem 2:

a = np.array([6, 4, 2, 9, 5, 1, 8, 3, 7])

print(a)

print(a) # prints [6 4 2 9 5 1 8 3 7]
print(a[:]) # prints [6 4 2 9 5 1 8 3 7]
print(a[4:]) # prints [5 1 8 3 7]
print(a[:4]) # prints [6 4 2 9]
print(a[3:8]) # prints [9 5 1 8 3]

print(a[5 :- 2]) # prints [1 8]
print(a[-2:3]) # prints []
print(a[-2:3 :- 1]) # prints [3 8 1 5]

print(a[ ::- 1]) # prints [7 3 8 1 5 9 2 4 6]
print(a[4 :: 2]) # prints [5 8 7]

print(a[:3 :- 2]) # prints [7 8 5]
print(a[:2 :- 2]) # prints [7 8 5]

print(a[-1 ::- 2]) # prints [7 8 5 2 6]
print(a[-1:0 :- 2]) # prints [7 8 5 2]

[6 4 2 9 5 1 8 3 7]
[6 4 2 9 5 1 8 3 7]
[6 4 2 9 5 1 8 3 7]
[5 1 8 3 7]
[6 4 2 9]
[9 5 1 8 3]
[1 8]
[]
[3 8 1 5]
[7 3 8 1 5 9 2 4 6]
[5 8 7]
[7 8 5]
[7 8 5]
[7 8 5 2 6]
[7 8 5 2]


In [ ]:
# Problem 3

import numpy as np

a = np.array([[0, 1, 2, 3],
[4,15, 6, 7],
[8, 9,1,2],
[3, 4, 5, 0]])

print(a[:, 1]) # second column
# [1 5 9 4]

print(a[2, :]) # third row
# [89 1 2]

print(a[3, ::- 1]) # fourth row backwards
# [0 5 4 3]

########### a[row_start:row_end, col_start:col_end] ###############

# m=row, n=column
# a[ :- m, :- n] # Exclude last m rows and n columns

print(a[ :- 2, :- 2]) # all lines and columns except last two
# [[0 1]
#[4 5]]

# a[-m: ,-n: ] take only last m rows and n columns
print(a[-2:, -2:]) # all lines and columns except first two
# [[1 2]
# [5 0]]

print(a[-1 :- 2 :- 1]) # last row
# [[3 4 5 0]]

print(a[ :- 2, -2:]) # top two rows intersected with last two columns
# [[2 3]
# [6 7]]

print(a[1:3,1:3]) # middle two by two matrix
# [[5 6]
# [9 1]]

print(a[ ::- 1, ::- 1]) # reversed rows in reverse
# [[0 5 4 3]
# [219 8]
# [7 6 5 4]
# [321 0]]

[ 1 15  9  4]
[8 9 1 2]
[0 5 4 3]
[[ 0  1]
 [ 4 15]]
[[1 2]
 [5 0]]
[[3 4 5 0]]
[[2 3]
 [6 7]]
[[15  6]
 [ 9  1]]
[[ 0  5  4  3]
 [ 2  1  9  8]
 [ 7  6 15  4]
 [ 3  2  1  0]]


Broadcasting simplifies mathematical operations on arrays with different shapes. It enable
NumPy to efficiently apply operations element-wise without explicitly copying or reshaping
data.

In [ ]:
# Problem 4

a = np.array([
[1, 2, 0],
[4, 5, 3],
[6, 8,7]])

a[ :: 2] += 1 # broadcasting happens here
print (a) # prints [[2 3 1] [[4 5 3] [7 9 8]]
# rows 1 and 3 have their elements incremented by one

print("----------")

a[:, :: 2] += 1 # broadcasting, again

print(a) # prints [[3 3 2] [5 5 4] [8 9 9]]
# columns 1 and 3 have their elements incremented by one

print("----------")

a[:, : ] -= 1 # what is broadcasting?
print(a) # prints [[2 2 1] [4 4 3] [7 8 8]]
# all elements decremented by one

print("----------")
print(a.shape) # prints (3, 3)
print(a[1].shape) # prints (3,)
print(a.ndim) # prints 2
print(a[0].ndim) # prints 1


[[2 3 1]
 [4 5 3]
 [7 9 8]]
----------
[[3 3 2]
 [5 5 4]
 [8 9 9]]
----------
[[2 2 1]
 [4 4 3]
 [7 8 8]]
----------
(3, 3)
(3,)
2
1


In [ ]:
# Problem 5

a = np.array([
              [0, 1, 0],
              [2, 0, 0],
              [3, 0, 3]
                        ])

print(np.nonzero(a)) # prints (array([0, 1, 2, 2]), array([1, 0, 0, 2]))
print(np.average(a)) # prints 1.0 (9/9)
print(a > np.average(a)) # prints [[False False False] [True False False] [ True False  True]]

print("----------------------------------")

b = np.array(["One", "Two", "Three"])

print(np.nonzero(a > np.average(a))) # (array([1, 2, 2]), array([0, 0, 2]))
print(np.nonzero(a > np.average(a))[0]) # [1 2 2]

print(b[ np.nonzero(a > np.average(a)) [0] ]) # ['Two' 'Three' 'Three' ]
print(set(b[ np.nonzero(a > np.average(a)) [0] ])) # {'Two', 'Three'}

(array([0, 1, 2, 2]), array([1, 0, 0, 2]))
1.0
[[False False False]
 [ True False False]
 [ True False  True]]
----------------------------------
(array([1, 2, 2]), array([0, 0, 2]))
[1 2 2]
['Two' 'Three' 'Three']
{'Three', 'Two'}


In [ ]:
# Problem 6

a = np.array([[25, "mgbako"],
              [15, "mgbako" ],
              [16, "reneau"],
              [19, "reneau"],
              [16, "goode"],
              [15, "gallowa]y"],
              [15, "reneau"]      ])

# pu away
# wi away
# mi at home
# msu away
# pu home
# ucla home
# pu home

## One-liner
b = a[a[:,0].astype(float) > a[:,0].astype(float).min(), 1]

# a[: ,0].astype(float) -> selects all the rows but only 0th column and then conver
# a[:,0].astype(float).min() -> gives 15.
# Filtering Rows Where Column 0 Is Greater Than the Minimum -> a[:,0].astype(float
# then it selects only from the column 1 and stores the names

## Results
print(b) # prints ['mgbako' 'reneau' 'reneau' 'goode' ]
print(set(b)) # prints {'mgbako', 'reneau', 'goode'}

['mgbako' 'reneau' 'reneau' 'goode']
{'reneau', 'goode', 'mgbako'}


In [ ]:
# Problem 7

a = np.array([1, 2, 3, 4, 5, 6])

a[1 :: 2] = 1 # a = [1,1,3,1,5,1] # Shape (6,)
print(a.reshape(2, -1)) # prints [[1 1 3] [1 5 1]]

print("---------------------------------")

a[2 :: 3] = np.sum(a.reshape((-1,3)), axis=1)
print(a) # prints [1 1 5 1 5 7]
# replace every third value with the sum of the last three values

print("---------------------------------")

points = np.array([15, 15, 16, 19, 16, 15, 25])
players = np.array(["reneau", "galloway", "goode", "reneau", "reneau", "mgbako", "mgbako"])

print(np.sort(points)) # prints [15 15 15 16 16 19 25]
print(np.argsort(points)) # prints [0 1 5 2 4 3 6]

# players[np.argsort(points)] > Rearrange players in the same sorted order.
print(players[np.argsort(points)][0:3]) # prints ['reneau' 'galloway' 'mgbako' ]

# prints names of bottom three top scorers

print("---------------------------------")

numbers = np.array([5, 1, 4, 6, 2, 3])
numbers = numbers.reshape(-1, 3)

print(numbers) # prints [[5 1 4] [6 2 3]]
print(max(numbers, key = lambda x : x[1])) # prints [6 2 3] # Finds the row with t
# from the last problem in chapter 3

[[1 1 3]
 [1 5 1]]
---------------------------------
[1 1 5 1 5 7]
---------------------------------
[15 15 15 16 16 19 25]
[0 1 5 2 4 3 6]
['reneau' 'galloway' 'mgbako']
---------------------------------
[[5 1 4]
 [6 2 3]]
[6 2 3]


In [ ]:
# Problem 8

import re

# re.sub(pattern, replacement, input)

input = "I am happy."

print(re.sub('[aeiou]', '-', input)) # prints I -m h-ppy.

print(re.sub('[^aeiou]', '-', input) ) # prints -- a --- a ----

print(re.sub('([^aeiou\s.])',r'(\1)', input)) # prints (I) a(m) (h)a(p)(p)(y).
# [^aeiou\s.] > Matches any character that is NOT a vowel, space, or period.
# () around \1 in replacement > Wraps the matched character inside parentheses.

print(re.sub('(\W)(\w)',r'\2\1', input)) # prints Ia mh appy.
# \W > Matches any non-word character (spaces, punctuation)
# \w > Matches any word character (letters, digits)
# r'\2\1' > Swaps positions of the matched groups.

print(re.sub('(\w)(\W)',r'\2\1', input)) # prints Ia mhapp.y

print(re.sub('([aeiou])([^aeiou\s])', r'\2\1', input)) # prints I ma hpapy.
# ([aeiou]) > Matches a vowel (a, e, i, o, u).
# ([^aeiou\s]) > Matches any consonant, but not spaces.

I -m h-ppy.
--a---a----
(I) a(m) (h)a(p)(p)(y).
Ia mh appy.
 Ia mhapp.y
I ma hpapy.


In [ ]:
# Problem 9

input = "Once 9123 upon a 4567 there was a 12389"
print(input) # prints Once 9123 upon a 4567 there was a 12389

result = re.sub('\d\d', '-', input)
print(result) # prints Once -- upon a -- there was a -- 9

# Pattern: \d((\d)(\d))\d
# \d > Matches a digit before and after the captured group.
# ((\d)(\d)) > Captures two digits inside a bigger group.
# \1 First captured group (second and third digit).
# \3 Third captured group (third digit).

result = re.sub('\d((\d)(\d))\d', r'\1+\3', input)
# This pattern replaces a string with a part of itself
print(result) # prints Once 12+2 upon a 56+6 there was a 23+39

julianne = "2364523"
result = re.sub('(\d)(\d)', r'\1+\2', julianne)
print(result) # prints 2+36+45+23

def add_two(match):
  return str(int(match.group()) + 2)

s = 'Alex 28'
result = re.sub(r'(\d+)', add_two, s)
print(result) # prints Alex 30

# \d+ > Matches any sequence of digits (28)

suraj = re.sub(r'(\d)(\d)', lambda m: str(int(m.group(1))+int(m.group(2))), s)
print(suraj) # prints Alex 10

# Pattern: (\d)(\d)
# Matches two consecutive digits.
# Replacement: Adds them together.

linjun = "12345"
suraj = re.sub(r'(\d)(\d)', lambda m: m.group(2) +m.group(1), linjun)

print(suraj) # prints 21435

# same as above just reverse the order

text = "I am here."
text = re.sub('(\w)', r'(\1)', text)
print(text) # prints (I) (a)(m) (h)(e)(r)(e).

Once 9123 upon a 4567 there was a 12389
Once -- upon a -- there was a --9
Once 12+2 upon a 56+6 there was a 23+39
2+36+45+23
Alex 30
Alex 10
21435
(I) (a)(m) (h)(e)(r)(e).


In [ ]:
# Problem 10

def nuf(a, b):
  return int(a) + int(b)

def fun(match): # , eggplant):
  a = nuf(match.group(1), match.group(2))
  print(match) # what gets printed with this?
  return str(a) # match.group(2) + match.group(1)

word = "12345"
result = re.sub('(\d)(\d)', fun, word)
# (\d) (\d) > Captures two consecutive digits.
# 1,2 got matched and 3,4 got matched which returned 3,7 from the function
print(result) # prints 375

mark = map(lambda num : num + 1, [8, 1, 2, 3, 2, 5]) # just adds one to each element
print(list(mark)) # prints [9, 2, 3, 4, 3, 6]

word = "The works."
result1 = re.sub('(\w)',r'(\1)', word)
result2 = re.sub('(\w)','(\1)', word)
#'(\1)' (without r'') is treated as a normal string, not a raw string.

print(result1) # prints (T)(h)(e) (w)(o)(r)(k)(s).
print(result2) # prints (T)(h)(e) (w)(o)(r)(k)(s).

<re.Match object; span=(0, 2), match='12'>
<re.Match object; span=(2, 4), match='34'>
375
[9, 2, 3, 4, 3, 6]
(T)(h)(e) (w)(o)(r)(k)(s).
()()() ()()()()().


In [ ]:
# Problem 11

from sklearn.linear_model import LinearRegression
import numpy as np

## Data (Apple stock prices)
apple = np.array([157, 156, 159, 159])
n = len(apple)

## One-liner
model = LinearRegression().fit(np.arange(n).reshape((n,1)), apple)

# np.arange(n).reshape((n,1))
# Becomes:
#[[0]
# [1]
# [2]
# [3]]
# X = [[ø], [1], [2], [3]] (time steps)
# y = [157, 156, 159, 159] (stock prices)

# y = mx+c formula

## Result & puzzle
print(model.predict([[4], [5], [6]]) )

[160.  160.9 161.8]


In [ ]:
# Problem 12

from sklearn.linear_model import LogisticRegression

## Data (#cigarettes, cancer)
x = np.array([[0, "No"],
[10, "No"],
[60, "Yes"],
[90, "Yes"]])

## One-liner
model = LogisticRegression().fit(x[:,0].reshape(n,1), x[:,1])

# X[:,0] do? Selects only the first column (number of cigarettes).
# .reshape(n,1)?
# X[:,0] gives a 1D array, but sklearn expects a 2D array (n,1).
# reshape(n,1) ensures it is correctly formatted.

# then x and y fits into logistic regression which works on sigmoid function

## Result & puzzle
print(model.predict([[12], [29], [34], [36], [38], [41], [58]]))

['No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'Yes']



Supervised machine learning can be separated into two phases:

1) The training phase: you giving an x input and the model predicts output y and compares it to y', which is your desired output.
2) The inference phase: you use the trained model to predict output values for new input features x.

---

Unsupervised machine learning helps to model the unlabeled data when it’s not clear what “the optimal output” means. However, you can still distill useful knowledge from these unlabeled data sets (for example, to find clusters of similar unlabeled data).

---

Linear regression: the one machine learning algorithm for machine learning. Every machine learning model consists of model parameters. Model parameters are internal configuration variables that are estimated from the data. These model parameters determine how exactly the model calculates the prediction, given the input features.

---

Logistic regression is commonly used for classification problems, in which you predict whether a sample belongs to a specific category (or class). It
fits an S-shaped curve, called the sigmoid function. The S-shaped curve helps you make binary decisions (for example, yes/no).

---

K-means clustering is an unsupervised learning technique that groups similar data points together, where k is the number of groups to be identified. It is a simple, efficient, and popular way of extracting k clusters from multidimensional data.

---

K-nearest neighbors is a supervised learning technique that classifies a data point by referencing the classifications of other data points it is closest to, where k is the number of data points to reference. It is used for regression and classification in many applications such as recommender systems, image classification, and financial data forecasting.

---

An artificial neural network is used to break the big task of learning and inference into multiple micro-tasks. These micro-tasks are not independent but interdependent. The idea is that it can associate images with their corresponding labels and recognize the new and unabled 'values'.

---

Neural network analysis mean for example the first five questions will be the input, and the sixth question will be the output. In this one-liner section,
You predict a numerical value based on numerical input features.

---

Decision trees are human-readable. It is extremely easy to explain to someone who has no background in machine learning. They represent a structured way of making decisions. Each decision opens new branches. By answering a bunch of questions, you’ll land on the recommended outcome.

---

Variance measures the expected (squared) deviation of the data from its mean. It captures the patterns of the data in a generalized manner—and machine learning is all about pattern recognition.  It’s a simple statistical property that captures how much the data set spreads from its mean.

---

Basic statistics: knowledge of basic statistics helps extracting variables/information from matrices (such as average, variance, and standard deviation). It is a critical component for analyzing a wide range of data sets such as financial data, health data, or social media data.

---

A support-vector machine can work even if there are more dimensions/features than data items. This is unusual for classification algorithms because of the curse of dimensionality: with increasing dimensionality, the data becomes extremely sparse, which makes it hard for algorithms to find patterns
in the data set.

---

Random forests are supervised learning technique that generates multiple decision trees by selecting different combinations of binary questions at random to grow each tree branch.
Good for predictions from
individual trees are then aggregated to generate final predictions.
Each decision tree is built by injecting randomness in the tree-generation procedure during the training
phase. So you can decide which tree node to select first.

---

A regular expression formally describes a search pattern that can be used to match sections of text. The most basic regular expression is a simple
string pattern, such as a word/string 'Juliet'.
It is good for finding arbitrary text patterns in strings or developing a web scraper which regularly can pull the HTML source code of  websites, or matching a single-digit number. Basically, retrieve a certain string that you want from the texts.

---

Some supervised (machine) learning algorithms:
linear regression, logistic regression, decision trees,
random forest, support-vector machines, neural networks, and k-nearest neighbors.

---

Some unsupervised (machine) learning algorithms:
to calculate basic statistics of multidimensional data
arrays, and to use the K-Means algorithm.

---

I have read about association rules in "Numsense!" chapters 2 and 4.
An unsupervised model could learn which items were
frequently bought together using association rules, explained in Chapter 4, or it could cluster customers based on their purchases explained in Chapter 2.

Association Rules. An unsupervised learning technique that dis-
covers how data points are associated with each other, such as identifying items that are frequently bought together.

---

Same book mentions "social network analysis" in chapter 5.
"To examine how these people relate to each other, such as identifying prominent individuals and how they drive group dynamics, we could use a technique called Social Network Analysis (SNA)" (42).

---

A/B testing is for example displaying different version of an ad to 100 people in a pilot trial, to see how many eventually clicked on each ad. The ad that garnered more clicks would appear to attract
more buyers, and thus should be used for the rest of advertising campaign. Basically, the comparison of the
effectiveness of ad versions A and B. A/B testing has some limitations such as the result can ba a fluke and potential loss of revenue.

Multi-arm bandits are process of choosing which slot machine to play. A term that now refers to any problem on resource allocation, such as deciding which ad to show, which topics to revise before an exam, or which drug study to fund.

---

What is PCA (Principal Component Analysis)
An unsupervised learning technique that reduces the number of variables we have to analyze by combining the most informative variables in our data into new variables called principal components.”
NumSense! Summary: “allows us to express our data with a smaller set of variables known as principal components. Hence PCA is called a dimension-reduction technique.”


------
More Detailed Explanations below
----

Define: supervised machine learning.
Look at page 121 in NumSense! Glossary and you will see: “A class of machine learning algorithms that is used when we want to make predictions. These algorithms are supervised because we want them to base their predictions on pre-existing patterns on our data.” See also page 7 that says: “task: use the patterns in my data to make predictions”. Also look on page 82 in Python One-Liners where we distinguish: training phase and inference phase.

----

Define: unsupervised machine learning.
From Glossary p. 122 NumSense! (see also questions 15 and 16 in this list): “A class of
machine learning algorithms that is used when we want to find hidden patterns in our data.
These algorithms are unsupervised because we do not know what patterns to look out for
and thus leave them to be uncovered by the algorithms. See also p. 6 in NumSense! that
summarizes this as “task: tell me what patterns exist in my data.”

----

What is linear regression?
See Regression Analysis p. 120 Glossary in NumSense! that says: “A supervised learning
technique that finds the best fit trend line that passes through or sits close to as many data
points as possible. The trend line is derived from a weighted combination of predictors.”
See also chapter 2 in Dive into Data Science (we will go over soon in lab, has excellent
examples). See also page 83 in Python One-Liners that says: “commonly used in regression
problems for which the model predicts missing data values by using existing ones.”

----

What is logistic regression?
See p. 100 in Python One-Liners that says: “linear and logistic regression algorithms define
learning parameters, while the clustering algorithm (in this case it’s about the k-Nearest
Neighbors in One Line example) calculates cluster centers based on the training data.” See
also p. 126 in Python One-Liners that says: “a classification algorithm (compare this with
the prediction made by linear regression)”. A point we make here is with respect to the two
algorithms mentioned in the two questions that follow. One of them (k-means clustering)
is an unsupervised learning technique (or, algorithm), the other one (k-nearest neighbors,
as mentioned already in this paragraph in the quote) is supervised learning.

---

What is k-means clustering?
See p. 119 Glossary NumSense! “An unsupervised learning technique that groups similar
data points together, where k is the number of groups to be identified.” See also p. 94 in
Python One-Liners (the section that presents k-Means Clustering in One Line).

----

What is k-nearest neighbors?
See p. 119 Glossary NumSense! “A supervised learning technique that classifies a data
point by referencing the classifications of other data points it is closest to, where k is the
number of data points to reference.” See also p. 100 in Python One-Liners. See also the
earlier question [ 4] in this list.

---

What is an artificial neural network?
See p. 100 in section11.6 Summary of NumSense! “A neural network comprises layers of
neurons. During training, neurons in the first layer are activated by input data, and these
activations are propagated to neurons in subsequent layers, eventually reaching the final
output layer to make predictions.”
Meaningful terms to look up: activation rule, feedback on prediction accuracy (back
propagation), and supervised learning. “Neural networks perform best when large data sets
and advanced computing hardware are available. Results would nonetheless be largely
uninterpretable.” See also p. 106 in Python One-Liners.

---

What does neural network analysis mean?
A supervised learning technique that uses layers of neurons to transmit activations for
learning and making predictions. While highly accurate, results are largely uninterpretable
due to its complexity. (From pp. 119-120, Glossary in NumSense!) See also page 104 in
Python One-Liners (and thereabouts).

---

What are decision trees?
See p. 118 Glossary in NumSense! “A supervised learning technique that makes predictions
by asking a sequence of binary questions to repeatedly partition data points to obtain
homogeneous groups. While easy to understand and visualize, decision trees are prone to
overfitting.” Look for this supervised learning technique in Dive into Data Science (coming
soon to a lab near you this semester). See also p. 111 in Python One-Liners and thereabouts.

---

What is variance?
Explained in Python One-Liners on p. 114 (or thereabouts). It’s a simple statistical property
that captures how much the data set spreads from its mean. The variance is the sum of the
squared deviations to the mean (this describes the actual formula, which is in the book).

---

Define basic statistics.
See p. 117 (and around it) in Python One-Liners. First define simple average, then based
on it, define variance (for that, see previous question, too). Then the square root of the
variance is called the standard deviation.


---

What is a support-vector machine?
See p. 119 in Python One-Liners: “SVMs have gained massive popularity in recent years
because they have robust classification performance, even in high-dimensional spaces.”
On p. 120 “ how do [they] work? They use the training data to find a decision boundary
that divides data in the one class from data in the other class.” See also page 121 in the
Glossary NumSense! “A supervised learning technique that classifies data points into two
groups by drawing a boundary down the middle between the peripheral data points, also
called support vectors, of both groups. It employs the kernel trick to efficiently derive
curved boundaries.” Look up “the kernel trick” too.

----

What are random forests and what are they good for?
See p. 120 Glossary NumSense! “A supervised learning technique that generates multiple
decision trees by selecting different combinations of binary questions at random to grow
each tree branch. Predictions from individual trees are then aggregated to generate final
predictions.” Look for random forests also in Python One-Liners, Dive into Data Science.

----

What is a regular expression? What is it good for?
See p. 161 for chapter 7 in Automate the Boring Stuff with Python (2nd edition). See also
p. 126 for chapter 5 in Python One-Liners. If you google it the dictionary entry is: regular
expression, noun (computing) “a sequence of symbols and characters expressing a string
or pattern to be searched for within a longer piece of text.” See also Wikipedia page Stephen
Kleen in 1951 introducing them for (or via) regular languages (first paragraph in History).

---

Mention some supervised (machine) learning algorithms.
See table 3 on page 6 in NumSense! where the following are listed: Regression Analysis,
k-Nearest Neighbors, Support Vector Machine, Decision Tree, Random Forests, Neural
Networks. Unsupervised Learning algorithms are also mentioned and there is a separate
category for Reinforcement Learning (Multi-Armed Bandits). See also Appendix B in
NumSense! on page 110. Note: prediction, analysis, results (are the rows in the table).

----

Mention some unsupervised (machine) learning algorithms.
See table 3 on page 6 in NumSense! where the following are listed: k-Means Clustering,
Principal Component Analysis, Association Rules, Social Network Analysis. There is also
a category for Reinforcement Learning (with Multi-Armed Bandits as the only algorithm
listed in it) and Supervised Learning algorithms are listed there as well.
In Python One-Liners page 94 we have k-Means Clustering in One Line (but that’s about
the only one time in that book) and also a summary on page 126. See also Appendix A on
page 109 NumSense! where we note that the rows in the table are: input, and output.

---

Where have you read or heard about "association rules" in this class?
See Glossary on page 116 in NumSense! that says: “An unsupervised learning technique
that discovers how data points are associated with each other, such as identifying items that
are frequently bought together. There are three common measures of association: support,
confidence and lift.” See also chapter 9 in Dive into Data Science (e.g., on page 191).

---

Same question about "social network analysis".
See page 52 in NumSense! that says: “Social Network Analysis allows us to map and
analyze relationships such as those between people, countries or other entities. The
Louvain method (look it up) identifies clusters in a network (in a way that maximizes
interactions within clusters, and minimizes those between clusters) and is an Unsupervised
Learning Algorithm. Also unsupervised is the PageRank algorithm that ranks nodes in a
network based on the[ir] number of links, as well as the strength and source of those links.”

---

What is A/B testing? What are multi-arm bandits?
See Glossary NumSense! on p. 121 Reinforcement Learning: “a class of machine learning
algorithms that is used when we want to make predictions based on patterns in our data,
and to continuously improve those predictions as more results come in.”


Multi-arm bandit
problem: see page 119 in the Glossary that says “a term used to refer to any problem on
resource allocation, such as deciding which slot machine to place your bets on. The term
was inspired by the moniker for slot machines – one arm-bandits – as they appear to cheat
players of money with each arm pull.” See also p. 116 Glossary in NumSense! where the
definition for A/B Testing is: “a strategy to compare the returns from two products A and
B. The process starts with an exploration phase, where both products are tested at the same
rate. After which, the better product is identified and all resources are devoted to it to
maximize returns in an exploitation phase. The degree of trade-off between exploration (to
check for better alternatives) and exploitation (to increase potential rewards) is a central
decision for the conduct of any A/B test.” See also the whole chapter 4 in Dive into Data
Science and on page 108 in NumSense! the corresponding Summary.

---

What is PCA (Principal Component Analysis)
Unsupervised learning technique (see Appendix A on p. 109 in NumSense!). See also
Glossary on page 120 in NumSense! which says: “An unsupervised learning technique that
reduces the number of variables we have to analyze by combining the most informative
variables in our data into new variables called principal components.” See also p. 23 in
NumSense! Summary: “[…] allows us to express our data with a smaller set of variables
known as principal components. Hence PCA is called a dimension-reduction technique.”